lib/src/local_working_copy.rs

```diff
@@ -1457,9 +1457,9 @@ impl TreeState {
         Ok(())
     }
 
-    pub fn reset_to_empty(&mut self) {
+    pub fn recover(&mut self, new_tree_id: MergedTreeId) {
         self.file_states.clear();
-        self.tree_id = self.store.empty_merged_tree_id();
+        self.tree_id = new_tree_id;
```

Comment:
In `snapshot()`, `deleted_files` are detected by subtracting from the `file_states`. So resetting only the `tree_id` would break the assumption there.

If I changed the test as follows, the debug assertion failed.
```diff
diff --git a/cli/tests/test_workspaces.rs b/cli/tests/test_workspaces.rs
index edb85c84f4...3924bf144d 100644
--- a/cli/tests/test_workspaces.rs
+++ b/cli/tests/test_workspaces.rs
@@ -430,7 +430,8 @@
         &["sparse", "set", "--clear", "--add=file1"],
     );
     // Also leave an unsnapshotted write in the secondary working copy
-    std::fs::write(secondary_path.join("file1"), "secondary\n").unwrap();
+    //std::fs::write(secondary_path.join("file1"), "secondary\n").unwrap();
+    std::fs::remove_file(secondary_path.join("file1")).unwrap();
 
     // Create an op by abandoning the parent commit. Importantly, that commit also
     // changes the target tree in the secondary workspace.
```

Reply:
Good catch! Thanks for finding that. I've added some more tests and updated the code to basically first reset to the empty tree and then reset to the recovery commit.